### DISCUSSION TITLE
Operational & Functional View

### DISCUSSION THREAD

**Discussion by smcdonou02** [dyc3/sys-581-group-project#16] [General]
Come up with 2-3 Operational & Functional View of our system. Explain each with 2-3 sentences

**Comment by smcdonou02** [comment]
Operational View: 
[Earth User (NASA) - sends  message to system](https://drive.google.com/file/d/1bX0O09pCeIgl72AirUWUYozK4vbCbVrS/view?usp=sharing)

**Comment by dyc3** [comment]
```mermaid
sequenceDiagram
    actor Alice

    Alice-->>C3: Queues Message
    C3-->>Alice: Queued
    Note over C3: System passes message across relay nodes
    actor Bob
    C3-->>Bob: Flush Queue
    Bob-->>C3: Received
    C3-->>Alice: Ack Receipt
```

**Reply by nick-dimeglio** [reply]
This diagram showcases how Alice (user from earth) would use the system to send a message to Bob (user from Mars). Alice queues a message to be sent with our system, receives confirmation of it being queued, then the system passes the queue of messages through the appropriate relays to get the message to mars. When the queue reaches Mars, it is flushed and all messages (including the one from Alice) will be received by Bob. Once this occurs, a confirmation is sent back across the system to be acknowledge a successful message transfer to Alice.

**Comment by dyc3** [comment]
```mermaid
sequenceDiagram
    Relay1-->Relay2: Connection Established
    Relay1-->Relay2: Negotiate Transfer Speed
    par Data Transfer
    Relay2-->>Relay1: Flushes Queue
    and Data Transfer
    Relay1-->>Relay2: Flushes Queue
    end
    break when connection interrupted or degrades too far
    Note over Relay1,Relay2: Incomplete queues saved
    end
```

**Reply by nick-dimeglio** [reply]
This diagram goes into more detail about how relays communicate with each other within our system. Each relay will establish a connection with the desired next in line, then a simultaneous queue flush occurs where both message queues (in our case, one coming from Earth and one coming from Mars) are flushed and sent to the other relay. In the case that there is any error or connection is lost, both relays save the outstanding and incomplete queues and await for the next opportunity to pass the message along to the next relay.

**Comment by smcdonou02** [comment]
![image](https://github.com/dyc3/sys-581-group-project/assets/114617841/af2c472d-a519-4f49-94e6-2392830e6223)

This Operational View demonstrates data (video, text or image files) being passed from colonists on Mars to the C3 system, and then outputted data is received by NASA on Earth. Inputted data from Earth (message response in form of video, text or image files) is then sent back to the system and outputted to Mars colonist.

**Comment by smcdonou02** [comment]
Document compiled with both operational and functional view documents
[Google Doc](https://docs.google.com/document/d/1TE7ZY-NDTcY4mbEc6BMuJYZunXnOT7_R7CiGj4UMJVU/edit?usp=sharing)

**Comment by Victor-M-Giraldo** [comment]
```mermaid
flowchart TD
    A[C3 System]
    A --> B[Send message]
    A --> C[Receive message]
    A --> D[Confirm message sent]
    A --> E[Relay message to another node]
```

**Comment by dyc3** [comment]
```mermaid
stateDiagram-v2
    state "Connection Established" as established
    state "Flush Queue" as flush
    state "Receive Messages" as recv

    [*] --> established: Node makes contact
    established-->flush: Have messages destined for Node
    established-->[*]: No messages in queue,\nNone to receive
    established-->recv: Node has messages for us
    flush-->recv: Swap
    recv-->flush: Swap
    recv-->[*]
    flush-->[*]
```

**Reply by nick-dimeglio** [reply]
This functionally explains how a node communicates with another node in the system. It makes contact, establishes connection, then if it has no messages to receive and no messages to send, nothing happens. If it has a message destined for the connected node, it will flush it's queue and send it over. However, if it has messages destined AND the node has messages destined for the node itself, it will simultaneously flush the queues to each other.
